### DISCUSSION TITLE
Alertifyjs implementation

### DISCUSSION THREAD

**Discussion by phibersoft** [vercel/next.js#12689] [Help]
Hi first,

I want to add alertifyjs to my project, but when I import it, it gives an error even though I don't use it. I think it's about server-side. But I couldn't find a resource I could solve / get ideas.

AlertifyJS : [https://alertifyjs.com/](https://alertifyjs.com/)

Code : 
`import alertify from "alertifyjs"`

Error : 
`
ReferenceError: document is not defined
(anonymous function)
[PATH]\node_modules\alertifyjs\build\alertify.js:293:13
(anonymous function)
[PATH]\node_modules\alertifyjs\build\alertify.js:304:6
(anonymous function)
[PATH]/node_modules/alertifyjs/build/alertify.js:3660:3
Module._compile
internal/modules/cjs/loader.js:1156:30
Module._extensions..js
internal/modules/cjs/loader.js:1176:10
Module.load
internal/modules/cjs/loader.js:1000:32
Function.Module._load
internal/modules/cjs/loader.js:899:14
Module.require
internal/modules/cjs/loader.js:1042:19
require
internal/modules/cjs/helpers.js:77:18
alertifyjs
webpack:/external "alertifyjs":1

> 1 | module.exports = require("alertifyjs");

__webpack_require__
./webpack/bootstrap:21

  18 | // Execute the module function
  19 | var threw = true;
  20 | try {
> 21 | 	modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
     | ^  22 | 	threw = false;
  23 | } finally {
  24 | 	if(threw) delete installedModules[moduleId];

Module../components/containers/AddUserForm.js
/_next/development/server/static/development/pages/create-user.js:1516:68
__webpack_require__
./webpack/bootstrap:21

  18 | // Execute the module function
  19 | var threw = true;
  20 | try {
> 21 | 	modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
     | ^  22 | 	threw = false;
  23 | } finally {
  24 | 	if(threw) delete installedModules[moduleId];

Module../pages/create-user.js
/_next/development/server/static/development/pages/create-user.js:3591:92
__webpack_require__
./webpack/bootstrap:21

  18 | // Execute the module function
  19 | var threw = true;
  20 | try {
> 21 | 	modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
     | ^  22 | 	threw = false;
  23 | } finally {
  24 | 	if(threw) delete installedModules[moduleId];

4
/_next/development/server/static/development/pages/create-user.js:3644:18
__webpack_require__
./webpack/bootstrap:21

  18 | // Execute the module function
  19 | var threw = true;
  20 | try {
> 21 | 	modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
     | ^  22 | 	threw = false;
  23 | } finally {
  24 | 	if(threw) delete installedModules[moduleId];

This screen is visible only in development. It will not appear if the app crashes in production.
Open your browser’s developer console to further inspect this error.`

**Comment by Geowan** [comment]
You might need to change how you load alertifyjs. I do it this way


            if (typeof window !== 'undefined') {
                  const alertify = require('alertifyjs');
                  alertify.confirm('Confirmation','Are you sure you want to do this?',  () => {
                               // do stuff on confirmation
                        },  ()=> {
                           alertify.error('Cancelled');
                   });
           }

**Comment by icyJoseph** [comment]
Looks like that library accesses document, at the top level scope of a given module. So simply importing it, causes the server side runtime, to execute that top level scope. This is how JavaScript modules work.

You might want to use next/dynamic, or a more generic dynamic import solution, to skip loading that module during server side.

Alternatively, it that's an open source library, maybe look into helping with a refactor, so that top level access to document is not needed.
